demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1685 @@
+ */
+static void mqttEventCallback( MQTTContext_t * pMqttContext,
+                               MQTTPacketInfo_t * pPacketInfo,
+                               MQTTDeserializedInfo_t * pDeserializedInfo );
+
+/**
+ * @brief Callback registered with the OTA library that notifies the OTA agent
+ * of an incoming PUBLISH containing a job document.
+ *
+ * @param[in] pContext MQTT context which stores the connection.
+ * @param[in] pPublishInfo MQTT packet information which stores details of the
+ * job document.
+ */
+static void mqttJobCallback( MQTTContext_t * pContext,
+                             MQTTPublishInfo_t * pPublishInfo );
+
+/**
+ * @brief Callback that notifies the OTA library when a data block is received.
+ *
+ * @param[in] pContext MQTT context which stores the connection.
+ * @param[in] pPublishInfo MQTT packet that stores the information of the file block.
+ */
+static void mqttDataCallback( MQTTContext_t * pContext,
+                              MQTTPublishInfo_t * pPublishInfo );
+
+static SubscriptionManagerCallback_t otaMessageCallback[ OtaNumOfMessageType ] = { mqttJobCallback, mqttDataCallback };
+
+/*-----------------------------------------------------------*/
+
+void otaEventBufferFree( OtaEventData_t * const pxBuffer )
+{
+    if( sem_trywait( &bufferSemaphore ) == 0 )
+    {
+        pxBuffer->bufferUsed = false;
+        ( void ) sem_post( &bufferSemaphore );
+    }
+    else
+    {
+        LogError( ( "Failed to get buffer semaphore: "
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+OtaEventData_t * otaEventBufferGet( void )
+{
+    uint32_t ulIndex = 0;
+    OtaEventData_t * pFreeBuffer = NULL;
+
+    if( sem_trywait( &bufferSemaphore ) == 0 )
+    {
+        for( ulIndex = 0; ulIndex < otaconfigMAX_NUM_OTA_DATA_BUFFERS; ulIndex++ )
+        {
+            if( eventBuffer[ ulIndex ].bufferUsed == false )
+            {
+                eventBuffer[ ulIndex ].bufferUsed = true;
+                pFreeBuffer = &eventBuffer[ ulIndex ];
+                break;
+            }
+        }
+
+        ( void ) sem_post( &bufferSemaphore );
+    }
+    else
+    {
+        LogError( ( "Failed to get buffer semaphore: "
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+
+    return pFreeBuffer;
+}
+
+/*-----------------------------------------------------------*/
+
+static void otaAppCallback( OtaJobEvent_t event,
+                            const void * pData )
+{
+    OtaErr_t err = OtaErrUninitialized;
+
+    /* OTA job is completed. so delete the MQTT and network connection. */
+    if( event == OtaJobEventActivate )
+    {
+        LogInfo( ( "Received OtaJobEventActivate callback from OTA Agent." ) );
+
+        if( pthread_mutex_lock( &mqttMutex ) == 0 )
+        {
+            /* OTA job is completed. so delete the network connection. */
```

Comment:
This doesn't delete the network connection thought, only MQTT. Is Openssl_Disconnect() supposed to get called after this?

Reply:
Openssl_Disconnect() is called in the `disconnect()` after we are done with the demo loop.

Reply:
`disconnect()` also calls `MQTT_Disconnect()`. I don't think there should be multiple calls to mqtt disconnect

Reply:
457df0c